test/app-luatest/gh_7746_fiber_channel_close_test.lua

```diff
@@ -0,0 +1,113 @@
+local compat = require('tarantool').compat
+local t = require('luatest')
+local g = t.group()
+local fiber = require('fiber')
+
+g.test_channel_close_default = function()
+    -- No graceful close (old) by default.
+    local ch = fiber.channel(10)
+    ch:put('one')
+    ch:put('two')
+    t.assert_equals(ch:get(), 'one')
+    ch:close()
+    t.assert(ch:is_closed())
+    t.assert_not(ch:put('three'))
+    t.assert_equals(ch:get(), nil)
+end
+
+g.test_channel_close_old = function()
+    compat.fiber_channel_graceful_close = 'old'
+    local ch = fiber.channel(10)
+    ch:put('one')
+    ch:put('two')
+    t.assert_equals(ch:get(), 'one')
+    ch:close()
+    t.assert(ch:is_closed())
+    t.assert_not(ch:put('three'))
+    t.assert_equals(ch:get(), nil)
+end
+
+g.test_channel_close_new = function()
+    -- Graceful close (new) selected.
+    compat.fiber_channel_graceful_close = 'new'
+    local ch = fiber.channel(10)
+    ch:put('one')
+    ch:put('two')
+    ch:close()
+
+    t.assert(ch:is_closed())
+    t.assert_not(ch:put('three'))
+
+    t.assert_equals(ch:get(), 'one')
+    t.assert(ch:is_closed())
+
+    t.assert_equals(ch:get(), 'two')
+    t.assert(ch:is_closed())
+
+    t.assert_equals(ch:get(), nil)
+end
+
+g.test_close_graceful_on_empty = function()
+    compat.fiber_channel_graceful_close = 'new'
+    local ch = fiber.channel(10)
+    ch:put('one')
+    ch:get()
+    ch:close()
+    t.assert(ch:is_closed())
+end
+
+g.test_close_reader_old = function()
+    compat.fiber_channel_graceful_close = 'old'
+    local ch = fiber.channel(0)
+    local writer = fiber.create(function()
+        t.assert_not(ch:is_closed())
+        t.assert_equals(ch:get(), nil)
+        t.assert(ch:is_closed())
+        t.assert_equals(ch:get(0), nil)
+    end)
+    ch:close()
+    t.assert(ch:is_closed())
+    fiber.yield()
+end
```

Comment:
There are several problems in this test case.

1. Typo: a reader fiber is named `writer`.
2. Assertions inside the fiber don't affect the test result. Let's check it:

   ```lua
   local fiber = require('fiber')

   local t = require('luatest')
   local g = t.group()

   g.test_x = function()
       fiber.create(function()
           t.assert_equals(1, 2)
       end)
   end
   ```
3. `fiber.yield()` in the main fiber doesn't guarantee that the reader fiber will be executed. Let's wait for this event in a loop.
4. There are many assertions and it is not clear on the first glance what is the point of the test. Let's make it explicit somehow: either adjust the code or add a comment.
5. There is no difference in old/new behavior for this situation. It would be nice to highlight this fact to don't let anyone compare those two test cases line-by-line.

Reply:
Fixed everything mentioned above.